sdk/javascript/wasm/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+
+use sha2::{Digest as Sha2Digest, Sha512};
+use sha3::{Digest as Sha3Digest, Keccak512};
+
+use js_sys::Uint8Array;
+use wasm_bindgen::prelude::*;
+
+mod utils;
+use crate::utils::set_panic_hook;
+
+/// Length of an ed25519 secret key, in bytes.
+pub const SECRET_KEY_LENGTH: usize = 32;
+
+/// Length of an ed25519 public key, in bytes.
+pub const PUBLIC_KEY_LENGTH: usize = 32;
+
+/// Length of a private key hash, in bytes.
+pub const HASH_LENGTH: usize = 64;
+
+/// Length of a signature, in bytes.
+pub const SIGNATURE_LENGTH: usize = 64;
+
+/// Supported hash modes.
+#[derive(Copy)]
+#[derive(Clone)]
+#[derive(PartialEq)]
+#[wasm_bindgen]
+pub enum HashMode {
+	/// Keccak hash.
+	Keccak,
+
+	/// SHA2 hash.
+	Sha2_512
+}
+
+// region Hasher512
+
+struct Hasher512 {
+	hash_mode: HashMode,
+	_keccak: Keccak512,
+	_sha2: Sha512
+}
+
+impl Hasher512 {
+	fn new(hash_mode: HashMode) -> Hasher512 {
+		return Hasher512{
+			hash_mode: hash_mode,
+			_keccak: Keccak512::new(),
+			_sha2: Sha512::new()
+		}
+	}
+
+	fn update<T: AsRef<[u8]> + ?Sized>(&mut self, input: &T) {
+		if HashMode::Keccak == self.hash_mode {
+			self._keccak.update(&input);
+		} else {
+			self._sha2.update(&input);
+		}
+	}
+
+	fn finalize_into(self, out: &mut [u8; HASH_LENGTH]) {
+		if HashMode::Keccak == self.hash_mode {
+			out.copy_from_slice(&self._keccak.finalize().as_slice());
+		} else {
+			out.copy_from_slice(&self._sha2.finalize().as_slice());
+		}
+	}
+}
+
+// endregion
+
+// region helpers
+
+fn hash_private_key(hash_mode: HashMode, sk: &[u8; SECRET_KEY_LENGTH]) -> [u8; HASH_LENGTH] {
+	let mut hasher = Hasher512::new(hash_mode);
+	hasher.update(&sk);
+
+	let mut hash: [u8; HASH_LENGTH] = [0u8; HASH_LENGTH];
+	hasher.finalize_into(&mut hash);
+	hash
+}
+
+fn copy_and_clamp_private_hash(private_hash: &[u8; HASH_LENGTH]) -> Scalar {
+	let mut bits: [u8; SECRET_KEY_LENGTH] = [0; SECRET_KEY_LENGTH];
+	bits.copy_from_slice(&private_hash[..SECRET_KEY_LENGTH]);
+
+	bits[0] &= 248;
+	bits[31] &= 127;
+	bits[31] |= 64;
+
+	Scalar::from_bits(*&mut bits)
+}
+
+fn hash_to_scalar(builder: Hasher512) -> Scalar {
+	let mut hash = [0u8; HASH_LENGTH];
+	builder.finalize_into(&mut hash);
+	Scalar::from_bytes_mod_order_wide(&hash)
+}
+
+fn generate_nonce(hash_mode: HashMode, private_hash: &[u8; HASH_LENGTH], message: &[u8]) -> Scalar {
+	// hash half of priv key hash
+	let mut builder = Hasher512::new(hash_mode);
+	builder.update(&private_hash[32..64]);
+	builder.update(&message);
+
+	hash_to_scalar(builder)
+}
+
+// endregion
+
+// region implementations
+
+/// Generates public key given secret key.
+pub fn crypto_sign_keypair_unboxed(hash_mode: HashMode, sk: &[u8; SECRET_KEY_LENGTH]) -> [u8; PUBLIC_KEY_LENGTH] {
+	set_panic_hook();
+
+	let private_hash = hash_private_key(hash_mode, sk);
+	let key = copy_and_clamp_private_hash(&private_hash);
+	let point = &key * &curve25519_dalek::constants::ED25519_BASEPOINT_TABLE;
+	let compressed = point.compress();
+	compressed.to_bytes()
```

Comment:
I'm wondering if we should zero key and/or private_hash after calculating point here... (and similar thing in signing below, zero after calculating S)